src/webgpu/util/texture/layout.ts

```diff
@@ -0,0 +1,194 @@
+import * as C from '../../../common/constants.js';
+import { assert, unreachable } from '../../../common/framework/util/util.js';
+import { kTextureFormatInfo } from '../../capability_info.js';
+import { align, isAligned } from '../math.js';
+
+export const kBytesPerRowAlignment = 256;
+export const kBufferCopyAlignment = 4;
+
+export function getTextureCopyLayout(
+  format: GPUTextureFormat,
+  dimension: GPUTextureDimension,
+  size: [number, number, number],
+  mipLevel: number = 0,
+  bytesPerRow?: number,
+  rowsPerImage?: number
+): {
+  bytesPerBlock: number;
+  byteLength: number;
+  minBytesPerRow: number;
+  bytesPerRow: number;
+  rowsPerImage: number;
+} {
+  let mipSize: [number, number, number];
+  switch (dimension) {
+    case '1d':
+      mipSize = [size[0] >> mipLevel, size[1], size[2]];
+      break;
+    case '2d':
+      mipSize = [size[0] >> mipLevel, size[1] >> mipLevel, size[2]];
+      break;
+    case '3d':
+      mipSize = [size[0] >> mipLevel, size[1] >> mipLevel, size[2] >> mipLevel];
+      break;
+    default:
+      unreachable();
+  }
+
+  const { blockWidth, blockHeight, bytesPerBlock } = kTextureFormatInfo[format];
+  assert(!!bytesPerBlock && !!blockWidth && !!blockHeight);
```

Comment:
nit: `> 0`
although I don't think this assert is really necessary.

Reply:
They're `number | undefined` so it's both a type assertion and >0 check.